Split scrubbing into ./pkg/util/scrubber

djmitche opened the pull request:
This breaks the "scrubbing" functionality out of ./pkg/util/log and into its own package.  

It also renames things to consistently use "scrub" and its various conjugations.

### Motivation

Poorly-isolated functionality :)

### Additional Notes

This new package is a module, like ./pkg/util/log.  Mostly because ./pkg/util/log depends on it!

With this PR merged, I'd like to make a few followups:
 * Move the scrubbing functionality currently implemented for flares into the new ./pkg/util/scrubber package
 * Refactor scrubbing a bit to better define the "things" that can be scrubbed.  It's designed to scrub YAML files.  Its function names suggest it can scrub files, bytes, and URLs.  And in practice it's used to scrub strings that are none of those things (e.g., calling `ScrubURL(somethingThatIsNotAURL)`).
 * Clarify the distinction between scrubbing and the orchestrator's "redaction" support.  I think these are distinct, at least :)

### Describe how to test your changes

To test config scrubbing:
* Start an agent
* Run `agent config` and check that the API key is scrubbed

To test the `flare_stripped_keys` config:
* Set `flare_stripped_keys` to include a key in your config, e.g., `flare_stripped_keys: [logs_enabled]`
* Run `agent config` and check that the key is scrubbed

To test that logs are being scrubbed:
* Grep your agent log for `hasValidAPIKey` and see that it scrubbed the API key

djmitche committed:
Add ./pkg/util/cleaner

This is based on existing code in ./pkg/util/log

djmitche committed:
Global scrubber for pkg/util/log

djmitche committed:
Omit module pkg/util/logs from govet

Govet only wants to vet packages that are part of the current module,
which pkg/util/log is not.

This appears to have been an oversight -- the other modules (e.g.,
pkg/util/winutil) are already omitted.

djmitche committed:
Use cleaner for scrubbing

djmitche committed:
consistently refer to it as 'scrubbing'

djmitche committed:
fix go vet error

olivielpeau reviewed (commented):
If I understand the code correctly, a significant behavior difference this PR introduces is that `pkg/util/log` doesn't apply any scrubbing by default when imported as a module (~i.e. when imported from outside the datadog-agent repo).
I'm not sure we're comfortable with that behavioral change though. Typically I believe the OTLP collector's DD exporter imports this module cc @mx-psi .

Maybe we'd simply want to define `pkg/util/scrubber` as its own module to avoid the "funny business" you mention in the code? WDYT @djmitche ?

djmitche committed:
Make pkg/util/scrubber a module

Because pkg/util/log is a module, and that module depends on
pkg/util/scrubber, the latter must also be a module.

This includes some compatibility warnings in anticipation of
API-breaking changes in the future.

djmitche committed:
fix up module-ization

djmitche committed:
add indirect dependency to DEFAULT_MODULES

djmitche committed:
try referring to version 0.0.0

djmitche committed:
more tidy fixes

olivielpeau reviewed (approved):
LGTM once my comment is addressed. Please update the PR description as well (it currently reflects the PR's state before you made the new package a module).

djmitche committed:
include accidentally-commented code

djmitche committed:
Merge remote-tracking branch 'upstream/main' into dustin.mitchell/ac-1307-pkg-util-scrubber

mfpierre reviewed (approved):
LGTM for @DataDog/container-app
